tasks/rare-numbers/src/main.rs

```diff
@@ -0,0 +1,333 @@
+        .collect::<Vec<(u8, u8)>>();
+
+    let bp_odd = (0_u8..=9)
+        .cartesian_product(0_u8..=9)
+        .filter(|pair| (pair.0 as i32 - pair.1 as i32).abs() % 2 == 1)
+        .collect::<Vec<(u8, u8)>>();
+
+    let bp_9 = (0_u8..=9)
+        .cartesian_product(0_u8..=9)
+        .filter(|pair| pair.0 + pair.1 == 9)
+        .collect::<Vec<(u8, u8)>>();
+
+    let bp_73 = (0_u8..=9)
+        .cartesian_product(0_u8..=9)
+        .filter(|pair| [7, 3].contains(&(pair.0 as i8 - pair.1 as i8)))
+        .collect::<Vec<(u8, u8)>>();
+
+    let bp_11 = (0_u8..=9)
+        .cartesian_product(0_u8..=9)
+        .filter(|pair| pair.0 + pair.1 == 11 || pair.1 + pair.0 == 1)
+        .collect::<Vec<(u8, u8)>>();
+
+    let aq_bp_setup: Vec<((u8, u8), &Vec<(u8, u8)>)> = vec![
+        ((2, 2), &bp_equal),
+        ((4, 0), &bp_zero_or_even),
+        ((6, 0), &bp_odd),
+        ((6, 5), &bp_odd),
+        ((8, 2), &bp_9),
+        ((8, 3), &bp_73),
+        ((8, 7), &bp_11),
+        ((8, 8), &bp_equal),
+    ];
+
+    //generate AB-PQ combinations
+    let aq_bp = aq_bp_setup
+        .iter()
+        .map(|e| {
+            e.1.iter().fold(vec![], |mut out, b| {
+                out.push(vec![e.0 .0, b.0, b.1, e.0 .1]);
+                out
+            })
+        })
+        .flatten()
+        .collect::<Vec<_>>();
+
+    let mut results: Vec<RareResults> = Vec::new();
+    let mut counter = 0_u32;
+    let start_time = Instant::now();
+
+    for d in 1..=digits {
+        print!("Digits: {} ", d);
+
+        if d < 4 {
+            for n in 10_u64.pow((d - 1).into())..10_u64.pow(d.into()) {
+                if is_rare(n) {
+                    counter += 1;
+                    results.push(RareResults {
+                        digits: d,
+                        time_to_find: start_time.elapsed().as_millis(),
+                        counter,
+                        number: n,
+                    });
+                }
+            }
+        } else {
+            aq_bp.iter().for_each(|abqp| {
+                let start = abqp[0] as u64 * 10_u64.pow((d - 1).into())
+                    + abqp[1] as u64 * 10_u64.pow((d - 2).into())
+                    + 10_u64 * abqp[2] as u64
+                    + abqp[3] as u64;
+
+                // brute-force checking all numbers which matches the pattern AB...PQ
+                // very slow
+                for n in (start..start + 10_u64.pow((d - 2).into())).step_by(100) {
+                    if is_rare(n) {
+                        counter += 1;
+                        results.push(RareResults {
+                            digits: d,
+                            time_to_find: start_time.elapsed().as_millis(),
+                            counter,
+                            number: n,
+                        });
+                    }
+                }
+            });
+        }
+
+        println!(
+            "Digits: {} done - Elapsed time(ms): {}",
```

Comment:
I agree that it might not be the best idea to add the benchmarking considering the long run time. The original task is just to find the rare numbers, the timing is just an interesting aspect, as some of the calculations are CPU consuming, and with increasing # of digits its exponentially more time consuming.

Reply:
Bu I'll add a few tests!